### DISCUSSION TITLE
Smart typed environment variables

### DISCUSSION THREAD

**Discussion by phifa** [epicweb-dev/epic-stack#62] [Ideas]
So instead of doing this: `process.env.IS_NICE`

we could do `constants.IS_NICE`. 

Advantages: 

- typed 🚀 parsing the above with zod could make this a boolean 
- auto-complete, just type `constants.` and see what's available 
- Error on app start, if neccessary constants are not set. No more digging for why the app failed, when only an env var was missing. 

This could be a simple file, or two for like `constants.client.ts` and `constants.server.ts`.

**Comment by stephen776** [comment]
It’s nice to be able to still set ENV via `process.env`. it would be great to validate at runtime via zod to get the static typing and nice imports

**Reply by kentcdodds** [reply]
We already have validation of environment variables with TypeScript types and autocomplete. We don't need Zod for this.

**Reply by stephen776** [reply]
> We already have validation of environment variables with TypeScript types and autocomplete. We don't need Zod for this.

Can the current setup ensure that, for example, if I have an ENV variable that is intended to be a `number` and I pass a `string` at runtime via process.env that this will be caught?

**Reply by kentcdodds** [reply]
No, it doesn't do that. That would be useful. I'd be interested in a pull request that switches from the current validation to Zod parsing.

**Reply by stephen776** [reply]
Cool. I'll try to put something together.

**Reply by stephen776** [reply]
@kentcdodds What do we think of an api along these lines? I am working on some helpers for my current project with this idea in mind

![image](https://github.com/epicweb-dev/epic-stack/assets/565167/d107b5e6-e301-4cd8-97dc-c24b4744b519)

**Reply by kentcdodds** [reply]
I don't think we need an "API" really. I'd prefer to keep it as simple as `z.object({/* stuff */}).parse(process.env)`

**Reply by stephen776** [reply]
Ya - that's fair. I was toying with the idea of creating an npm package that would handle this sort of dev friendly way to define and consume ENV vars in remix - including safe handling of sending variables down to the client so that's why my head went to designing some sort of nice interface.

**Reply by sergiodxa** [reply]
> I don't think we need an "API" really. I'd prefer to keep it as simple as `z.object({/* stuff */}).parse(process.env)`

This is what I do too (I even have an article about that), there's no need for a package because at the end you will still need the Zod schema or shape from the developer anyway.

What I have been doing lately is to do that parse in the HTTP server, then use getLoadContext to pass env variables to the loaders/actions. So instead of doing `process.env.FOO` or `env.FOO` (where `env` is the result of Zod parse) I do `context.env.FOO`.

This way you also ensure the only way to access env variables is inside loaders and actions, and if you need them somewhere else you need to pass return them from the loader/action.

**Reply by kentcdodds** [reply]
Hmmmm... I don't think I want to have to pass env vars to functions. Thanks for the suggestion though!
